aggregator/src/main/java/com/redhat/cloud/notifications/db/EmailAggregationRepository.java

```diff
@@ -23,13 +24,15 @@ public class EmailAggregationRepository {
     EntityManager entityManager;
 
     public List<AggregationCommand> getApplicationsWithPendingAggregationAccordinfOrgPref(LocalDateTime end) {
+        // Must takes every EmailAggregation supposed to be processed on last 15 min
+        // it covers cases when aggregation job may be run with few minutes late (ie: 05:01, 07,32)
         String query = "SELECT DISTINCT ea.orgId, ea.bundleName, ea.applicationName, acp.lastRun FROM EmailAggregation ea, AggregationOrgConfig acp WHERE " +
             "ea.orgId = acp.orgId AND ea.created > acp.lastRun AND ea.created <= :end " +
-            "AND hour(acp.scheduledExecutionTime) = :runningHour";
-
+            "AND :endTime >= acp.scheduledExecutionTime AND (:endTime - acp.scheduledExecutionTime) < CAST(:cutoff as LocalTime)";
```

Comment:
Is it possible to replace
```
(:endTime - acp.scheduledExecutionTime) < CAST(:cutoff as LocalTime)
```
with something based on Postgres's [extract(minute ...)](https://www.postgresql.org/docs/11/functions-datetime.html#FUNCTIONS-DATETIME-EXTRACT) function?

Something like this (untested):
```
EXTRACT(MINUTE FROM TIME :endTime) - EXTRACT(MINUTE FROM TIME acp.scheduledExecutionTime) < 15
```

Reply:
My previous suggestion was indeed not so good :smile:

What about this alternative?
```java
    public List<AggregationCommand> getApplicationsWithPendingAggregationAccordinfOrgPref(LocalDateTime now) {
        // Must takes every EmailAggregation supposed to be processed on last 15 min
        // it covers cases when aggregation job may be run with few minutes late (ie: 05:01, 07,32)
        String query = "SELECT DISTINCT ea.org_id, ea.bundle, ea.application, acp.last_run FROM email_aggregation ea, aggregation_org_config acp WHERE " +
            "ea.org_id = acp.org_id AND ea.created > acp.last_run AND ea.created <= :now " +
            // BETWEEN is inclusive so we need to subtract one millisecond from the right boundary.
            "AND  CAST(:now AS TIME) BETWEEN acp.scheduled_execution_time AND acp.scheduled_execution_time + (15 * INTERVAL '1 minute') - (1 * INTERVAL '1 microsecond')";
        Query hqlQuery = entityManager.createNativeQuery(query)
                .setParameter("now", now);

        List<Object[]> records = hqlQuery.getResultList();
        return records.stream()
                .map(emailAggregationRecord -> new AggregationCommand(
                    new EmailAggregationKey((String) emailAggregationRecord[0], (String) emailAggregationRecord[1], (String) emailAggregationRecord[2]),
                    ((Timestamp) emailAggregationRecord[3]).toLocalDateTime(),
                    now,
                    DAILY
                ))
                .collect(toList());
    }
```

Reply:
My only concern about the current query  is that it not easy to understand at first glance.

Reply:
I think what you did is fine 😃